src/main/java/org/checkerframework/specimin/StandardTypeRuleDependencyMap.java

```diff
@@ -314,4 +339,149 @@ public List<Node> getRelevantElements(Object resolved) {
 
     return elements;
   }
+
+  /**
+   * Given a type declaration, return all must implement methods to the result list. Must implement
+   * methods are methods that are abstract in JDK superclasses or non-default methods in JDK
+   * interfaces.
+   *
+   * @param typeDecl The type declaration to which the parents belong
+   */
+  private List<MethodDeclaration> getAllMustImplementMethods(TypeDeclaration<?> typeDecl) {
+    List<MethodDeclaration> methods = new ArrayList<>();
+    Set<MethodDeclaration> alreadyImplemented = new HashSet<>();
+
+    List<ClassOrInterfaceType> parents = new ArrayList<>();
+
+    if (typeDecl instanceof NodeWithExtends<?> withExtends) {
+      parents.addAll(withExtends.getExtendedTypes());
+    }
+    if (typeDecl instanceof NodeWithImplements<?> withImplements) {
+      parents.addAll(withImplements.getImplementedTypes());
+    }
+
+    getAllMustImplementMethodsImpl(parents, typeDecl, typeDecl, methods, alreadyImplemented);
+
+    methods.removeAll(alreadyImplemented);
+    return methods;
+  }
+
+  /**
+   * Helper method for getAllMustImplementMethods. This method recursively finds all must implement
+   * methods in the type declaration's parents and ancestors.
+   *
+   * @param parents The parents of the type declaration, i.e., the extended/implemented types
+   * @param originalTypeDecl The original type declaration
+   * @param typeDecl The current type declaration being processed
+   * @param result The result list
+   * @param alreadyImplemented A set of methods that are already implemented in a superclass or
+   *     interface
+   */
+  private void getAllMustImplementMethodsImpl(
+      List<ClassOrInterfaceType> parents,
+      TypeDeclaration<?> originalTypeDecl,
+      TypeDeclaration<?> typeDecl,
+      List<MethodDeclaration> result,
+      Set<MethodDeclaration> alreadyImplemented) {
+    for (ClassOrInterfaceType type : parents) {
+      try {
+        ResolvedType resolved = type.resolve();
+
+        if (!resolved.isReferenceType()) {
+          continue;
+        }
+
+        boolean isInterface = false;
+        ResolvedReferenceTypeDeclaration resolvedTypeDecl =
+            resolved.asReferenceType().getTypeDeclaration().orElse(null);
+
+        Set<ResolvedMethodDeclaration> methods;
+
+        if (resolvedTypeDecl instanceof ReflectionClassDeclaration reflectionClassDeclaration) {
+          methods = reflectionClassDeclaration.getDeclaredMethods();
+        } else if (resolvedTypeDecl
+            instanceof ReflectionInterfaceDeclaration reflectionInterfaceDeclaration) {
+          isInterface = true;
+          methods = reflectionInterfaceDeclaration.getDeclaredMethods();
+        } else {
+          continue;
+        }
+
+        for (ResolvedMethodDeclaration resolvedMethodDecl : methods) {
+          // Don't preserve equals(Object) and hashCode()
```

Comment:
Why are these two methods, specifically, excluded? I would expect that any justification for excluding `equals` and `hashCode` would also apply to other object methods, like `clone`, no? If these two methods are special for some reason, this comment should explain why.

Reply:
> equals and hashCode are in the Set interface

But they are already defined on the object because they are defined in `java.lang.Object`, so being in an interface is immaterial. I think this exception should apply to all of the methods in `java.lang.Object`, not just to equals and hashCode. It's okay to just add a TODO though - I don't think this is very important.